src/downloader.rs

```diff
@@ -0,0 +1,127 @@
+    retries: u32,
+    /// Number of maximum concurrent downloads.
+    concurrent_downloads: usize,
+}
+
+impl Downloader {
+    /// Starts the downloads.
+    pub async fn download(&self, downloads: &[Download]) {
+        // Prepare the HTTP client.
+        let retry_policy = ExponentialBackoff::builder().build_with_max_retries(self.retries);
+        let client = ClientBuilder::new(reqwest::Client::new())
+            .with(TracingMiddleware)
+            .with(RetryTransientMiddleware::new_with_policy(retry_policy))
+            .build();
+
+        // Download the files asynchronously.
+        let _tasks = stream::iter(downloads)
+            .map(|d| self.fetch(&client, d))
+            .buffer_unordered(self.concurrent_downloads)
+            .collect::<Vec<_>>()
+            .await;
+    }
+
+    /// Fetches the files and write them to disk.
+    async fn fetch(&self, client: &ClientWithMiddleware, download: &Download) -> Result<(), Error> {
+        // Request the file.
+        debug!("Fetching {}", download.url.clone());
+        let res = client.get(download.url.clone()).send().await.unwrap();
+
+        // Prepare the destination directory/file.
+        fs::create_dir_all(&self.directory)?;
+        let output = self.directory.join(&download.filename);
+        debug!("Creating file {:?}", &output);
+        let mut file = File::create(output)?;
+
+        // Download the file chunk by chunk.
+        debug!("Retrieving chunks...");
+        let mut stream = res.bytes_stream();
+        while let Some(item) = stream.next().await {
+            let chunk = item?;
+            file.write_all(&chunk)?;
+        }
+
+        Ok(())
+    }
+}
+
+/// A builder used to create a [`Downloader`].
+pub struct DownloaderBuilder(Downloader);
+
+impl DownloaderBuilder {
+    /// Creates a builder with the default options.
+    pub fn new() -> Self {
+        DownloaderBuilder::default()
+    }
+
+    /// Sets the directory where to store the [`Download`]s.
+    pub fn directory(mut self, directory: PathBuf) -> Self {
+        self.0.directory = directory;
+        self
+    }
+
+    /// Set the number of retries per [`Download`].
+    pub fn retries(mut self, retries: u32) -> Self {
+        self.0.retries = retries;
+        self
+    }
+
+    /// Set the number of concurrent [`Download`]s.
+    pub fn concurrent_downloads(mut self, concurrent_downloads: usize) -> Self {
+        self.0.concurrent_downloads = concurrent_downloads;
+        self
+    }
+
+    /// Create the [`Downloader`] with the specified options.
+    pub fn build(self) -> Downloader {
+        Downloader {
+            directory: self.0.directory,
+            retries: self.0.retries,
+            concurrent_downloads: self.0.concurrent_downloads,
+        }
+    }
+}
+
+impl Default for DownloaderBuilder {
+    fn default() -> Self {
+        Self(Downloader {
+            directory: std::env::current_dir().unwrap_or_else(|_| std::path::PathBuf::from("")),
```

Comment:
[`unwrap_for_default`](https://doc.rust-lang.org/std/option/enum.Option.html#method.unwrap_or_default)

Reply:
Much nicer 👍

Reply:
Ahhh, clippy is not happy with that! 

```
error: use of `unwrap_or` followed by a function call
Error:    --> src/downloader.rs:110:48
    |
110 |             directory: std::env::current_dir().unwrap_or(std::path::PathBuf::from("")),
    |                                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try this: `unwrap_or_else(|_| std::path::PathBuf::from(""))`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#or_fun_call
```

I think that's why I used `unwrap_or_else` in the first place, but I had forgotten.

Reply:
Hmm it looks like you used `unwrap_or` rather than `unwrap_or_default`